### PR TITLE
adds `url.URL` wrapper and other improvements

### LINKED ISSUE #46: `url.URL` wrapper and enhancements
# Proposed Changes

- Adds a robust `url.URL` wrapper that handles all parsing issues and most of the edgecases
- Now any special characters(even the reserved RFC chars) are not encoded >https://github.com/projectdiscovery/nuclei/issues/3166
- Adds helper methods and more

## Use case

- All encoding issues are now abstracted
- integrity of nuclei templates is not compromised

## References

- https://github.com/projectdiscovery/nuclei/issues/3208
- https://github.com/projectdiscovery/nuclei/issues/3167
and more

### PR DESCRIPTION
**Pull request by tarunKoyalwar**
## Proposed Changes

closes #46

### CODE REVIEW

**Review by xm1k3 — APPROVED** [review]
lgtm

**Comment by tarunKoyalwar** [comment]
@Ice3man543  , changed PR to draft until PR for nuclei is ready

**Comment by tarunKoyalwar** [comment]
seems stable nothing to improve 
Ref: https://github.com/projectdiscovery/nuclei/pull/3211

File: url/merge_test.go [hunk 1]
```diff
@@ -58,9 +58,12 @@ func TestMergeUnsafePaths(t *testing.T) {
 			t.Errorf(err.Error())
 			continue
 		}
-		rurl.MergePath(v.Path2, true)
-		if rurl.Path != v.Expected {
-			t.Errorf("expected %v but got %v", v.Expected, rurl.Path)
+		err = rurl.MergePath(v.Path2, true)
+		if err != nil {
```
Comment (Mzack9999) [inline url/merge_test.go:62]: As in other packages we used `require` package. What do you think of using:
```go
require.Nil()
require.Equalf()
```

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
lgtm - suggesting `require` to make tests more concise

**Comment by tarunKoyalwar** [comment]
## Changes

- bug fix where `localhost` was not considered as valid hostname

**Review by Mzack9999 — APPROVED** [review]

**Pull request merged**